src/OrbitGl/MultivariateTimeSeries.h

```diff
@@ -26,83 +28,117 @@ class MultivariateTimeSeries {
+      typename std::map<uint64_t, std::array<double, Dimension>>::const_iterator;
 
   struct Range {
     TimeSeriesEntryIter start_inclusive;
     TimeSeriesEntryIter end_inclusive;
   };
   // If there is no overlap between time range [min_time, max_time] and [StartTimeInNs(),
-  // EndTimeInNs()], return std::nullopt. Otherwise return a range of entries affected by the time
+  // EndTimeInNs()], return empty array. Otherwise return a range of entries affected by the time
   // range [min_time, max_time] where:
-  // * `Range::inclusive` points to the entry with the time key right before the time range
+  // * the first entry with the time key right before the time range
   // (min_time, max_time) if exists; otherwise points to the fist entry.
-  // * `Range::inclusive` points to the entry with the time key right after the time range
+  // * the last entry with the time key right after the time range
   // (min_time, max_time) if exists; otherwise points to the last entry.
-  [[nodiscard]] std::optional<Range> GetEntriesAffectedByTimeRange(uint64_t min_time,
-                                                                   uint64_t max_time) const {
-    if (IsEmpty() || min_time >= max_time || min_time >= time_to_series_values_.rbegin()->first ||
+  [[nodiscard]] std::vector<std::pair<uint64_t, std::array<double, Dimension>>>
+  GetEntriesAffectedByTimeRange(uint64_t min_time, uint64_t max_time) const {
+    absl::MutexLock lock(&mutex_);
+    if (time_to_series_values_.empty() || min_time >= max_time ||
+        min_time >= time_to_series_values_.rbegin()->first ||
         max_time <= time_to_series_values_.begin()->first) {
-      return std::nullopt;
+      return {};
     }
 
-    auto first_iterator = GetPreviousOrFirstEntry(min_time);
-    auto last_iterator = GetNextOrLastEntry(max_time);
-    return Range{first_iterator, last_iterator};
+    auto current_it = GetPreviousOrFirstEntryIterator(min_time);
+    auto last_iterator = GetNextOrLastEntryIterator(max_time);
+
+    std::vector<std::pair<uint64_t, std::array<double, Dimension>>> result;
+    result.push_back(*current_it);
+    do {
+      ++current_it;
+      result.push_back(*current_it);
+    } while (current_it != last_iterator);
+
+    return result;
   }
 
  private:
+  [[nodiscard]] TimeSeriesEntryIter GetPreviousOrFirstEntryIterator(uint64_t time) const {
+    CHECK(!time_to_series_values_.empty());
+
+    auto iterator_lower = time_to_series_values_.upper_bound(time);
+    if (iterator_lower != time_to_series_values_.begin()) --iterator_lower;
+    return iterator_lower;
+  }
+
+  [[nodiscard]] TimeSeriesEntryIter GetNextOrLastEntryIterator(uint64_t time) const {
+    CHECK(!time_to_series_values_.empty());
+
+    auto iterator_higher = time_to_series_values_.lower_bound(time);
+    if (iterator_higher == time_to_series_values_.end()) --iterator_higher;
+    return iterator_higher;
+  }
+
   void UpdateMinAndMax(double value) {
     max_ = std::max(max_, value);
```

Comment:
Why the compiler doesn't complain about the lack of mutex of this method? Because it's always called by some already guarded method?

Reply:
It is always called by guarded code, but I am not sure if compiler will complain once -Werror=thread-safety is enabled.
looping in @beckerhe

Reply:
Yes, the compiler will complain, but you can annotate the method to inform the callers that calling this method requires the mutex to be acquired.

If we merge this PR before the thread-safety PR I will take care of it there. You could already add the annotations right now but you can't really test it, so I advice to wait.

Reply:
ok, I will merge this PR now.

Reply:
Acvtually I take it back - I annotated the methods.